man/POSIX/qalter.1
.TH "qalter" 1
.SH "NAME"
qalter \- change properties of jobs already submitted to the queue manager
.SH "SYNOPSIS"
.B qalter 
.I [<options>] <jobid> [ <jobid> ... ]
.SH "DESCRIPTION"
.PP
.B qalter
changes properties of jobs already submitted to the queue manager.

Note: Currently no properties may be changed while a job is running.
.SH "OPTIONS"
.TP
.B \-d \-\-debug
Turn on communication debugging
.TP
.B \-\-help
Displays the usage and a brief descriptions of the options
.TP
.B \-\-version
Displays client revision and Cobalt version
.TP
.B \-A \-\-project "project"
Associate the job with the allocation for project project. This is
used to properly account for machine usage. 
.TP
.B \-\-attrs <attributes>
Modify attributes for jobs in args. Format (attr1=val1:attr2=val2:...:attrN=valN)'
.TP
.B \-\-dependencies
for jobs in args (jobid1:jobid2:..)
.TP
.B \-\-disable_preboot
If --mode script is specified, this will cause Cobalt to run the job script without
booting the block first.  By default the script is executed after the requested block is booted.
.TP
.B \-\-enable_preboot
If prebooting a block has been disabled, re-enable it.
.TP
.B \-\-defer
Defer jobs in the argument list for the current user.
.TP
.B \-e \-\-error "stderr file path"
Send job stderr to file specified.
.TP
.B \-\-geometry <geometry_spec>
Sets the geometry of the block that the jobs should run on.  This is in
the form of AxBxCxDxE.  This must be the geometry of the compute block.
This may not be a fraction of a block. Geometry may be unset if "None"
is passed as the geometry specification.
.TP
.B \-h \-\-held
hold jobs in args
.TP
.B \-M \-\-notify "email address"
Send an email notification at the start and stop of the job to the
specified email address.
.TP
.B \-\-mode "mode"
Either co (for coprocessor mode) or vn (for virtual node mode) or script (for script jobs)
.TP
.B \-n \-\-nodecount "node count" 
Specifies the node count for a job
.TP
.B \-o \-\-output "stdout file path"
Send job stdout to file specified.
.TP
.B \-\-proccount "process count"
Specify the number of processes to start
.TP
.B \-\-run_users \-\-user_list <user1>:<user2>
Sets a colon-separated list of users for the job being submitted.  All users
in this list will be able to execute cobalt commands to control the job. The 
submitting user is always able to run commands on a submitted job.
.TP
.B \-\-run_project
Appends the list of users in a project to the list of users authorized to run
commands on the submitted job. 
.TP
.B \-t \-\-time "wall clock time" 
Specify the runtime for a job. The time specified is the full amount in minutes
allotted to the job for execution. If the job runs over this limit, it
will be killed. If colon-delimited time is specified, the format is HH:MM:SS.

Also note that relative time is allowed.  A time argument with a '+' at the
beginning will add the argument amount to the current walltime, while an argument
with a '-' at the beginning will subtract the argument amount from the current
walltime.
.SH "SEE ALSO"
.BR qsub (1),
.BR qstat (1)
.SH "BUGS"


CHANGES
= Changes from previous Cobalt Versions =

== Changes to 0.99.33 ==
 * Fixed issue where adding blocks while the system component is running can
   cause the system component's block update loop to fail. (Cobalt Ticket #695)
 * Preliminary fix for a racecondition that can cause subblock jobs to terminate
   early (Cobalt Ticket #690)
 * Added T-Minus to showres (Cobalt Ticket #692)
 * Fixed an error in qalter where attemtpts to change the users on a running job
   caused qalter to fail. (Cobalt Ticket #694)

== Changes to 0.99.32 ==
 * Cluster systems now drain for larger jobs, and will backfill into the
   draining hardware. (Cobalt Ticket #663)
 * Fixed a bug in the prologue_helper where a silent failure on sending the
   hostfile to nodes could occur silently (Cobalt Ticket #686)

== Changes to 0.99.31 == 
 * Interactive mode jobs (qsub -I) are now supported on the BlueGene/Q.
   (Cobalt Ticket #107)
 * Timestamps are now included in cobaltlog lines (Cobalt Ticket #344 and #669)
 * Multiple mode arguments will result in a qsub rejection (Cobalt Ticket #664)
 * Fixed an issue that was causing COBALT_STARTTIME and COBALT_ENDTIME to not be
    set for compute node jobs in cX jobs. (Cobalt Ticket #673)

== Changes to 0.99.30 ==
 * This is a fix release that corrects a bug that can be encountered when upgrading
   to 0.99.27 from prior versions where ion_kerneloptions in cqm is improperly set.
   (Cobalt Ticket #665)

== Changes to 0.99.29 ==
 * Interactive jobs are experimentally available for cluster systems.

== Changes to 0.99.28 ==
 * Cobalt now supports qsub options in script mode jobs similarly to other
   job scheduler scripts.  The form is #Cobalt <qsub option> <option argument>.
   Most qsub options are supported, except for mode.  Jobs using this feature must
   be script mode jobs.  This is similar in style to PBS-directives. (Cobalt Ticket 354)
 * Fixed a bug in cluster systems where a failure duing job startup could result
   in all nodes for a job being stuck in the allocated state (Cobalt Ticket 662)
 * $COBALT_STARTTIME and $COBALT_ENDTIME added to job's environments, these are
   the Unix timestamp for when the job started, and when Cobalt will terminate
   the job due to exceeding walltime. (Cobat Ticket 593)
 * A requested walltime of 0 will set the job to run for the time remaining in
   a reservation. (Cobalt Ticket 474)
 * The --env flag to qsub now aggregates multiple instances of the flag rather
   than overriding the prior instance (Cobalt Ticket 657)

== Changes to 0.99.27 ==
 * Alternate kernel support for BG/Q compute nodes and IO nodes is available.
  Numerous configuration settings have been added to support this.
  (Cobalt ticket 621)
 * If $jobid is included in a job name specification or an environment variable,
  at submission time, Cobalt will substitute the generated jobid for that job.
  See man qsub for more details. (Cobalt ticket 617)
 * The Cobalt jobid is now explicitly added in the cobaltlog file. (Cobalt ticket 390)
 * The name of a job can now be specified seperately from naming the output
  files. (Cobalt ticket 625)
 * Multiple locations can be passed to reservations through the -p option as a
  ':'-delimited list. (Cobalt ticket 405)
 * cqadm can now be used to release user hold, as well as put jobs into a user hold
  as well as an admin hold (Cobalt ticket 382)
 * The users field can be removed from a reservation using setres -m -u '*'. (Cobalt
   ticket 381)
 * Releaseres should now report the correct partiton count on reservation release
  (Cobalt ticket 470)
 * Reservations now show time remaining in showres once the reservation is active.
  (Cobalt ticket 585)
 * Schedctl should handle flipping score and jobid arguments more gracefully and 
  no longer silently discards invalid options. (Cobalt ticket 502)
 * ':' and '=' may now be escaped as '\:' and '\=' in the --envs argument (Cobalt
  ticket 589)
 * Reservations may now take 'now' for the start time of a reservation (Cobalt
  ticket 546)
 * "GEOMETRY", "ION_KERNEL" and "ION_KERNELOPTIONS have been added to the JOB_DATA
  table of the Cobalt database 


== Changes to 0.99.26 ==
 * Corrected a bug that was causing messages in the clients that used to go to
  stdout to go to stderr instead

== Changes to 0.99.25 ==
 * Fixes for clients that were causing erronious default values for proccount
  have been fixed.
 * The setgid wrapper now calls a site-selectable python interperter with the
  '-E' flag by default.
 * A bug that could cause TimeRemaining to show negative values has been fixed
 * Queues may be added or removed from blocks without overwriting the entire
  list with new options in partadm using --rmq or --appq with the --queue option.


== Changes to 0.99.24 ==
 * Reverting client changes.  Found some issues when deploying to test systems
 * --defer flag has been reverted with these as a casualty.

== Changes to 0.99.23 ==
 * Client code has undergone a refactor.  Numerous bugs have been addressed
  with these that were found while improving test coverage of this code
 * IO Block booting and automatic control has been added.  Please consult the
  partadm manpage for details
 * This release requires pybgsched-0.1.3
 * A user may voluntarily set their job's score to 0 and let it re-accrue using
  qalter --defer (Ticket #590)

== Changes to 0.99.22 ==
 * Fix for a deadlock that could occur due to a race condition in the 
  boot-reaping code.
 * Fix for a pid collision that could occur after a forced delete of a job in
  the system-script-forker component leading to a job getting stuck in the
  Job_Epilogue state.
 * python 2.7 and it's change to the XMLRPC libraries are now supported.

== Changes to 0.99.21 ==
 * Minor fix release to correct a failure mode where we can lose a process group
  if a job's boot is delayed.

== Changes to 0.99.20 ==
 * Small changeset to fix a race condition that was hanging jobs that came in
   on 0.99.19

src/clients/POSIX/qalter.py
#!/usr/bin/env python
"""
Cobalt qalter command

Usage: %prog --help
Usage: %prog [options] <jobid1> ... <jobidN>
version: "%prog " + __revision__ + , Cobalt  + __version__

OPTIONS DEFINITIONS:

Option with no values:

'-d','--debug',dest='debug',help='turn non communication debugging',callback=cb_debug
'-v','--verbose',dest='verbose',help='not used',action='store_true'
'-h','--held',dest='user_hold',help='hold jobs in args',action='store_true'
'--run_project',dest='run_project',help='set run project flag for jogids in args',action='store_true'
'--enable_preboot',dest='script_preboot',help='enable script preboot',action='store_true'
'--disable_preboot',dest='script_preboot',help='enable script preboot',action='store_false'
'--defer',dest='defer', help='defer jobs in args for current user',action='store_true'

Option with values:

'-n','--nodecount',dest='nodes',type='int',help='modify job node count for jobs in args',callback=cb_nodes
'--proccount',dest='procs',type='int',help='modify job proc count for jobs in args',callback=cb_gtzero
'-A','--project',dest='project',type='string',help='modify project name for jobs in args'
'-M','--notify',dest='notify',type='string',help='modify notification email address for jobs in args'

'-t','--time',dest='walltime',type='string',help='modify walltime for jobs in args ([+/-]minutes or [+/-]HH:MM:SS', /
  callback=cb_time

'-e','--error',dest='errorpath',type='string',help='modify error file path for jobs in args'
'-o','--output',dest='outputpath',type='string',help='modify output file path for jobs in args'
'--dependencies',dest='all_dependencies',type='string',help='for jobs in args (jobid1:jobid2:..)',callback=cb_upd_dep
'--attrs',dest='attrs',type='string',help='modify attributes for jobs in args(attr1=val1:attr2=val2:...)',callback=cb_attrs
'--run_users',dest='user_list',type='string',help='modify user list for jobs in args (user1:user2:...)',callback=cb_user_list

The following optins are only valid on IBM BlueGene architecture platforms:

'--mode',dest='mode',type='string',help='modify system mode for jobs in args',callback=cb_mode
'--geometry',dest='geometry',type='string',help='modify geometry for jobs in args (AxBxCxDxE)',callback=cb_geometry

"""
import logging
import re
import sys
from Cobalt import client_utils
from Cobalt.client_utils import \
    cb_debug, cb_nodes, cb_time, cb_mode, \
    cb_upd_dep, cb_attrs, cb_user_list, cb_geometry, cb_gtzero
from Cobalt.Util import init_cobalt_config
from Cobalt.arg_parser import ArgParse

__revision__ = '$Revision: 559 $' # TBC may go away.
__version__ = '$Version$'

QUEMGR = client_utils.QUEMGR

init_cobalt_config()

def validate_args(parser, opt_count):
    """
    Validate qalter arguments
    """
    # Check if any altering options entered
    if opt_count == 0:
        client_utils.print_usage(parser, "No required options provided")
        sys.exit(1)

    # get jobids from the argument list
    jobids = client_utils.validate_jobid_args(parser)

    return jobids

def options_disallowed(parser):
    """
    This function will check the options against running jobs
    """
    _quit = False
    if parser.options.nodes != None:
        client_utils.logger.error( "cannot change node count of a running job")
        _quit = True
    if parser.options.procs != None:
        client_utils.logger.error( "cannot change processor count of a running job")
        _quit = True
    if parser.options.project != None:
        _quit = True
    if parser.options.notify != None:
        _quit = True
    if parser.options.mode != None:
        client_utils.logger.error( "cannot change mode of a running job")
        _quit = True
    if parser.options.walltime != None:
        client_utils.logger.error( "cannot change wall time of a running job")
        _quit = True
    if parser.options.errorpath != None:
        client_utils.logger.error( "cannot change the error path of a running job")
        _quit = True
    if parser.options.outputpath != None:
        client_utils.logger.error( "cannot change the output path of a running job")
        _quit = True
    if parser.options.all_dependencies != None:
        _quit = True
    if parser.options.attrs != None:
        client_utils.logger.error( "cannot change the attributes of a running job")
        _quit = True
    if parser.options.geometry != None:
        client_utils.logger.error( "cannot change the node geometry of a running job")
        _quit = True
    if parser.options.defer != None:
        client_utils.logger.error( "cannot change the score of a running job")
        _quit = True

    # Currently we do not support modifying a job while running
    if not _quit:
        client_utils.logger.error( "Modifying a job while running is currently not supported")
        _quit = True

    return _quit

def get_jobdata(jobids, parser, user):
    """
    Will get the jobdata for the specified jobs
    """

    jobs = [{'tag':'job', 'user':user, 'jobid':jobid, 'project':'*', 'notify':'*', 'walltime':'*',
        'mode':'*', 'procs':'*', 'nodes':'*', 'is_active':"*", 'queue':'*'} for jobid in jobids]

    jobdata = client_utils.component_call(QUEMGR, False, 'get_jobs', (jobs,))
    job_running = False

    # verify no job is running
    for job in jobdata:
        if job['is_active']:
            job_running = True

    if job_running:
        if options_disallowed(parser):
            sys.exit(1)

    return jobdata

def update_time(orig_job, new_spec, parser):
    """
    Update the new time. This will do delta time if specified.
    """
    minutes = new_spec['walltime']
    add_dt = False
    sub_dt = False

    if hasattr(parser.parser,'__timeop__'):
        if parser.parser.__timeop__ == '+': 
            add_dt = True
        elif parser.parser.__timeop__ == '-':
            sub_dt = True

    if add_dt:    # Add time to original job then
        new_spec['walltime'] = str(float(orig_job['walltime']) + float(minutes))

    elif sub_dt:  # Subtract time to original job and verify it is not less than 0
        new_time = float(orig_job['walltime']) - float(minutes)
        if new_time <= 0:
            client_utils.logger.error( "invalid wall time: " + str(new_time))
        else:
            new_spec['walltime'] = str(new_time)

    else:         # change to an absolute time
        new_spec['walltime'] = str(minutes)

def update_procs(spec, parser):
    """
    Will update 'proc' according to what system we are running on given the number of nodes
    """
    sysinfo = client_utils.system_info()
    if ((parser.options.nodes is not None or parser.options.mode is not None) and 
            parser.options.procs is None):
        if sysinfo[0] == 'bgq':
            if spec['mode'] == 'script':
                spec['procs'] = spec['nodes']
            else:
                rpn_re  = re.compile(r'c(?P<pos>[0-9]*)')
                mode_size = int(rpn_re.match(spec['mode']).group(1))
                spec['procs'] = spec['nodes'] * mode_size
        else:
            if parser.options.mode == 'vn':
                # set procs to 2 x nodes
                if sysinfo[0] == 'bgl':
                    spec['procs'] = 2 * spec['nodes']
                elif sysinfo[0] == 'bgp':
                    spec['procs'] = 4 * spec['nodes']
                else:
                    client_utils.logger.error("Unknown bgtype %s" % (sysinfo[0]))
                    sys.exit(1)
            else:
                spec['procs'] = spec['nodes']
    
def do_some_logging(job, orig_job, parser):
    """
    do some logging
    """
    # do some logging
    for key in job:
        if not orig_job.has_key(key):
            if key == "all_dependencies":
                if parser.options.all_dependencies != None:
                    client_utils.logger.info("dependencies set to %s" % ":".join(job[key]))
            else:
                client_utils.logger.info("%s set to %s" % (key, job[key]))
        elif job[key] != orig_job[key]:
            client_utils.logger.info("%s changed from %s to %s" % (key, orig_job[key], job[key]))

def main():
    """
    qalter main
    """
    # setup logging for client. The clients should call this before doing anything else.
    client_utils.setup_logging(logging.INFO)

    spec     = {} # map of destination option strings and parsed values
    opts     = {} # old map
    opt2spec = {}

    # list of callback with its arguments
    callbacks = [
        # <cb function>           <cb args>
        [ cb_debug               , () ],
        [ cb_gtzero              , (True,) ], # return int
        [ cb_nodes               , (True,) ], # return int
        [ cb_time                , (True, False, False) ], # delta time allowed, return minutes, return string
        [ cb_upd_dep             , () ],
        [ cb_attrs               , () ],
        [ cb_user_list           , (opts,) ],
        [ cb_geometry            , (opts,) ],
        [ cb_mode                , () ]]

    # Get the version information
    opt_def =  __doc__.replace('__revision__', __revision__)
    opt_def =  opt_def.replace('__version__', __version__)

    parser = ArgParse(opt_def, callbacks)

    user = client_utils.getuid()

    # Set required default values: None

    parser.parse_it() # parse the command line
    opt_count = client_utils.get_options(spec, opts, opt2spec, parser)

    # if run_project set then set run users to current user
    if parser.options.run_project != None:
        spec['user_list'] = [user]

    jobids  = validate_args(parser, opt_count)
    filters = client_utils.get_filters()

    jobdata = get_jobdata(jobids, parser, user)

    if parser.options.defer != None:
        client_utils.set_scores(0, jobids, user)
        if opt_count == 1:
            return

    response = []
    # for every job go update the spec info
    for job in jobdata:
        # append the parsed spec to the updates list
        new_spec          = spec.copy()
        new_spec['jobid'] = job['jobid']
        if parser.options.walltime is not None:
            update_time(job, new_spec, parser)
        if parser.options.nodes is not None or parser.options.mode is not None:
            if parser.options.nodes is None:
                new_spec['nodes'] = job['nodes']
            if parser.options.mode is None:
                new_spec['mode'] = job['mode']
            update_procs(new_spec, parser)
        if parser.options.geometry is not None:
            client_utils.validate_geometry(opts['geometry'], job['nodes'])

        del job['is_active']
        orig_job = job.copy()
        job.update(new_spec)

        client_utils.process_filters(filters, job)
        response = client_utils.component_call(QUEMGR, False, 'set_jobs', ([orig_job], job, user))
        do_some_logging(job, orig_job, parser)

    if not response:
        client_utils.logger.error("Failed to match any jobs or queues")
    else:
        client_utils.logger.debug(response)


if __name__ == '__main__':
    try:
        main()
    except Exception, e:
        client_utils.logger.fatal("*** FATAL EXCEPTION: %s ***", e)
        sys.exit(1)
